src/gccrs/args.rs

```diff
@@ -1,11 +1,85 @@
 //! This module interprets arguments given to `rustc` and transforms them into valid
 //! arguments for `gccrs`.
 
-use std::{env, path::PathBuf, process::Command};
+use std::convert::TryFrom;
+use std::ops::Deref;
+use std::path::{Path, PathBuf};
+
+use getopts::Matches;
+
+use super::{EnvArgs, Error, Result, RustcOptions};
+
+/// A collection containing multiple instances of `Options`. This is necessary in order
+/// to circumvent the fact that `rustc` can currently generate multiple types of binaries
+/// with a single invokation.
+///
+/// For example,
+/// `rustc --crate-type=static --crate-type=dyn --crate-type=bin src/<file>.rs` will attempt
+/// to generate an executable, a dynamic library and a static one.
+///
+/// When using `gccrs` to compile things, we must also take into account what `gcc` can
+/// and cannot do. And `gcc` cannot currently generate multiple outputs in a single invokation.
+///
+/// Let's look at the previous example:
+///
+/// * To generate an executable from `src/<file>.rs`,
+/// we'd need to do the following: `gccrs src/<file>.rs` (with `-o <file>` if we don't
+/// want an executable named a.out, but that's not important).
+///
+/// * For a shared library, we need to add the `-shared` flag. On top of this, `rustc`
+/// generates libraries named `lib<name>.[so|a]` on Linux, while `gcc` will happily generate
+/// a shared library without any extension or prefix. This amounts to the following command:
+/// `gccrs -shared src/<file>.rs -o lib<file>.so`.
+///
+/// * Finally, `gcc` is not able to generate a static library at all. We *need* to use a
+/// different command, `ar`, in order to bundle up object files previously created by
+/// `gcc`. Therefore, we actually need *two* commands:
+/// `gccrs -c src/<file>.rs && ar csr src/<file>.o`
+///
+/// Multiple flags, such as `shared`, conflict with the generation of other binaries. On
+/// top of that, we cannot use the `-o` option precisely enough to control the output name
+/// based on the binary file produced.
+///
+/// Therefore, we need to wrap an unknown amount of sets of `gccrs` arguments in order to
+/// mimic a single `rustc` invokation. Later on, we need to iterate over those sets and
+/// spawn a new `gccrs` command for each of them.
+pub struct OptionsCollection {
+    args_set: Vec<Options>,
+}
+
+/// Get the corresponding set of `gccrs` arguments from a single set of `rustc` arguments
+impl TryFrom<&[String]> for OptionsCollection {
+    type Error = Error;
 
-use getopts::{Matches, Options};
+    fn try_from(rustc_args: &[String]) -> Result<OptionsCollection> {
+        let matches = RustcOptions::new().parse(rustc_args)?;
```

Comment:
I would implement this on `RustcOptions` and not on `&[String]` directly.

Reply:
Fixed in e30a67f